pkg/domain/infra/abi/farm.go

```diff
@@ -0,0 +1,119 @@
+}
+
+func (ir *ImageEngine) fetchInfo(_ context.Context) (os, arch, variant string, nativePlatforms []string, emulatedPlatforms []string, err error) {
+	nativePlatform := parse.DefaultPlatform()
+	platform := strings.SplitN(nativePlatform, "/", 3)
+	switch len(platform) {
+	case 0, 1:
+		return "", "", "", nil, nil, fmt.Errorf("unparseable default platform %q", nativePlatform)
+	case 2:
+		os, arch = platform[0], platform[1]
+	case 3:
+		os, arch, variant = platform[0], platform[1], platform[2]
+	}
+	os, arch, variant = libimage.NormalizePlatform(os, arch, variant)
+	nativePlatform = os + "/" + arch
+	if variant != "" {
+		nativePlatform += ("/" + variant)
+	}
+	emulatedPlatforms = emulation.Registered()
+	return os, arch, variant, append([]string{}, nativePlatform), emulatedPlatforms, nil
+}
+
+// FarmNodeInspect returns information about the remote engines in the farm
+func (ir *ImageEngine) FarmNodeInspect(ctx context.Context) (*entities.FarmInspectReport, error) {
+	ir.platforms.Do(func() {
+		ir.os, ir.arch, ir.variant, ir.nativePlatforms, ir.emulatedPlatforms, ir.platformsErr = ir.fetchInfo(ctx)
+	})
+	return &entities.FarmInspectReport{NativePlatforms: ir.nativePlatforms,
+		EmulatedPlatforms: ir.emulatedPlatforms,
+		OS:                ir.os,
+		Arch:              ir.arch,
+		Variant:           ir.variant}, ir.platformsErr
+}
+
+// PullToFile pulls the image from the remote engine and saves it to a file,
+// returning a string-format reference which can be parsed by containers/image.
+func (ir *ImageEngine) PullToFile(ctx context.Context, options entities.PullToFileOptions) (reference string, err error) {
+	saveOptions := entities.ImageSaveOptions{
+		Format: options.SaveFormat,
+		Output: options.SaveFile,
+	}
+	if err := ir.Save(ctx, options.ImageID, nil, saveOptions); err != nil {
+		return "", fmt.Errorf("saving image %q: %w", options.ImageID, err)
+	}
+	return options.SaveFormat + ":" + options.SaveFile, nil
+}
+
+// PullToFile pulls the image from the remote engine and saves it to the local
+// engine passed in via options, returning a string-format reference which can
+// be parsed by containers/image.
+func (ir *ImageEngine) PullToLocal(ctx context.Context, options entities.PullToLocalOptions) (reference string, err error) {
+	destination := options.Destination
+
+	// already present at destination?
+	var br *entities.BoolReport
+	if destination == nil {
+		br, err = ir.Exists(ctx, options.ImageID)
+	} else {
+		br, err = destination.Exists(ctx, options.ImageID)
+	}
+	if err != nil {
+		return "", err
+	}
+	if br.Value {
+		return istorage.Transport.Name() + ":" + options.ImageID, nil
+	}
+
+	tempFile, err := os.CreateTemp("", "")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tempFile.Name())
+	defer tempFile.Close()
+
+	saveOptions := entities.ImageSaveOptions{
+		Format: options.SaveFormat,
+		Output: tempFile.Name(),
+	}
+	if err := ir.Save(ctx, options.ImageID, nil, saveOptions); err != nil {
```

Comment:
I am confused by this function.  Isn't `ir` the local runtime?  I'd expected us to use `destination` for saving the image and `ir` to load.

But the functions seems to support `destination == nil` in which case the image is saved and then re-load by the same runtime?

Reply:
Add an error check for `destination==nil` and fixed up the function to pull from `ir` and pull to `destination`.